### ISSUE TITLE
Bug: In Lighttable, selecting a range of images or selecting multiple images does not work

### ISSUE THREAD

**Issue by caigner** [darktable-org/darktable#11927]
I am on Gentoo Linux and am using Darktable v3.8.1-r1.

In Lighttable selecting multiple images by pressing **CTRL-left mouse click** and selecting a range of images by pressing **SHIFT-left mouse click** does not work.

Currently I can select multiple images only by hovering over images and pressing **space bar**.

Does anyone else have this problem or is it just me?

**Comment by Nilvus** [comment]
Probably just you. Such standard function would have been reported since 3.8.1 release few months ago. It works great on 3.8.1 and on master. I suggest you to check your shortcuts configuration. Maybe something broken on your side.
If you don't set customized shortcuts, first try to restore default shortcuts: click on keyboard icon on top right of thumbnails view, then click on a thumbnail for example to show shortcuts dialog window, then restore on bottom right. If that doesn't fix your issue, try to remove, with darktable closed, your darktablerc and shortcuts files in .config/darktable folder (save them of course to restore them if needed). I suggest to test that by removing on file first, then launch darktable and check what behavior you have.

Try also to check if you have same issue in other apps.

**Comment by caigner** [comment]
This really puzzles me. I did as you said. Deleted darktablerc and all shortcut files. Did not make any difference. I wonder, what could be the cause of this behaviour. In all other applications ctrl- and shift-left click works as expected. I welcome any input you can give me.

**Comment by Nilvus** [comment]
Really strange. I hope you remove both files with darktable closed. Those files are updated if darktable is launched during this operation.
Another thing, is Ctrl+a works to select all images?

If yes, I so can't update more here. Your issue is really strange and I don't have more ideas to help. Maybe someone else will see that and think about something else. As it works on darktable for most users, maybe try to ask also on discuss.pixls.us forum.

**Comment by caigner** [comment]
Yes, I removed both files while darktable was closed. CTRL-a works as expected.

I just tested selecting multiple entries in Thunderbird, and both CTRL-left click and SHIFT-left click work.
I also checked it in Nemo.

I think I will now downgrade Darktable and see if this makes a difference.

**Comment by caigner** [comment]
I just installed darktable v3.8.0 and I can select multiple images! Hurra!

Just for fun I shall upgrade to v3.8.1-r1 again.

**Comment by caigner** [comment]
I can confirm now that the problem is back again with version 3.8.1-r1 on Gentoo Linux.
So I guess I have to downgrade to 3.8.0 and live with it for a while.

**Comment by caigner** [comment]
To be precise: v3.8.0-r3 on Gentoo Linux works fine.

**Comment by Nilvus** [comment]
As 3.8.1 works great, there's probably an issue with 3.8.1 package.
3.8.1 fix a lot of bugs and was probably one of the top bugfixes release.
I so suggest you to send an issue to Gentoo maintainer of darktable package.

I so close as it's not a darktable issue but probably a package one.

**Comment by victoryforce** [comment]
darktable 3.8.1 on Ubuntu 22.04 (and Ubuntu 21.10 earlier): works fine.
darktable 3.8.1 on Windows 10: works fine.
darktable development builds on both OSes: works fine.

As we can easily see, the only constant factor that accompanies this problem is your Gentoo Linux distro. So I conclude that it would be best to address this issue to the Gentoo distro's darktable package maintainer. At least he/she works with this distribution and therefore can debug the issue if it's reproducible not only on your specific Gentoo installation.

**Comment by caigner** [comment]
>     * can you try to compile darktable master and see if your issue has been solved (would be by "accident", but...) Be careful to backup everything before (or better, use a "blank" config dir, just for the tests)

I compiled from this source: https://github.com/darktable-org/darktable/releases/download/release-3.8.1/darktable-3.8.1.tar.xz

I will try the master branch now and will report soon.

 
>     * if you still experiment the issue, may I propose you some line to add in the code, before recompiling, to try to find where the issue comes from...

I am not a C programmer, so I don't know, where I should add some text in the code.

**Comment by Nilvus** [comment]
I so reopen this one.

**Comment by rand00** [comment]
Copy paste from #12037 : 
I manually fixed another compilation problem on 3.8.0, and got it built.

The UI problems are not present in 3.8.0. So selection of multiple images works again + switch window focus doesn't stop shortcuts from working.

I'll try to bisect in the evening if I don't hit the same amount of compilation problems on other versions. Is there a flag I can pass to ignore minor warnings so I can be sure dt compiles while bisecting?

**Comment by AlicVB** [comment]
@rand00 : if you experiment problems with bisecting, you can also try the changes I have proposed to @caigner here : https://github.com/darktable-org/darktable/issues/11927#issuecomment-1154410551
They have to be applied on current master...
Thanks !

**Comment by rand00** [comment]
I tried your changes to master: 

* Shift-select works, but Ctrl select still doesn't. 
* Switch window focus back/forth also keeps shortcuts working now. 
* And dt now prints 'leave virtual' every time I left-click, no matter what key is pressed at the same time.

**Comment by AlicVB** [comment]
> * Shift-select works

Well I suppose that it doesn't works very well regarding the start of the selection... right ?



> * And dt now prints 'leave virtual' every time I left-click, no matter what key is pressed at the same time.

here is the issue I think... As I understand this, in your WM, for some reason, the leave event doesn't work the same than in "classic" WM...
first point will be to try to understand *how* it works... so I will need your help again !
remove the changes, and add `printf("leave detail=%d mode=%d\n", event->detail, event->mode);` in src/dtgtk/thumbtable.c in line 1043
Then I would be interested of the terminal output associated with the following action : 
1. move between thumbnails (try slow and quick movements from different parts of the thumb) => doesn't produce anything here
2. go to the end of the thumbnail list. You may have a "blank" area with no thumbnails at the bottom (or change the zoom level). what id the output if you move from thumbnail to this "blank" area
3. ensure you have some thumbnail "half shown" at the bottom. What is the output if you move from thumbnail to the bottom bar (quickly)
4. output when you click on a new thumbnail (try with ctrl or shift to see if that change anything)

That's all what I'm thinking of for instance... Thanks again for your help here !

**Comment by rand00** [comment]
> Well I suppose that it doesn't works very well regarding the start of the selection... right ?

I can select single images fine, and shift select works perfectly. Only Ctrl-select that doesn't work.

> first point will be to try to understand how it works... so I will need your help again !

Will try this out when I have some time - probably tomorrow. And thanks to you for fixing it (: I've used dt a lot for many years - fantastic software

**Comment by rand00** [comment]
1. Moving mouse between thumbnails, and on top of same thumbnail, doesn't trigger the print-event.
2. I get `leave detail=2 mode=0` when moving *in* to a thumbnail from the blank area. There is no printing when moving from thumbnail to blank area. 
3. With thumbnail half-down at the bottom
  * I get `leave detail=1 mode=0` when moving slowly
  * I get `leave detail=4 mode=0` when moving fast
4. When clicking new thumbnail
  * I get `leave detail=1 mode=0` when just left-clicking
  * The same event with both Ctrl/Shift click

**Comment by AlicVB** [comment]
Everything seems ok, except this : 

> * I get `leave detail=1 mode=0` when just left-clicking

which is very very suspicious : `detail = 1` means `GDK_NOTIFY_VIRTUAL`, which means, according to gdk doc : `the pointer moves between an ancestor and an inferior of the window.`
But here : the mouse is already inside an inferior of the thumbtable : a thumbnail. And when you "click" on it gtk seems to report that you leave this inferior for an ancestor of the thumbtable ??? Either there's something I don't understand in gtk code and docs, either your WM is doing something very strange...

I don't really know what to do here :/

Note that it's not the first time we have some strange events fired on some specific WM (fluxbox) : I remember we have had hard time to implement a workaround for something about double-click in lighttable, about 8 years ago... but when I have rewritten lighttable code, this was fixed upstream (sorry, I don't have the reference)

**Comment by rand00** [comment]
I agree that this behaviour seems weird, though I'm not experienced with GTK. Though from my point of view as a user:
* I don't have the same problems in other applications, of which many surely use GTK 
* and darktable used to work with my WM before a certain version

**Comment by victoryforce** [comment]
@caigner  If you confirm that the problem no longer exists, you can also close this issue.

**Comment by caigner** [comment]
As I opened this issue and the solution (whatever it was) works for me, I close it now.
Thanks to all who participated.

**Comment by rand00** [comment]
Yay :P I can also confirm it works
